File: app/src/main/java/io/jeffchang/buttondemo/ui/userlist/presenter/UserListPresenterImpl.java
package io.jeffchang.buttondemo.ui.userlist.presenter;

import io.jeffchang.buttondemo.network.remote.FakeButtonService;
import io.jeffchang.buttondemo.ui.userlist.view.UserListView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Implementation of dog list presenter.
 */

public class UserListPresenterImpl implements UserListPresenter {

    private final String CANDIDATE_ID = "vbnguirofbm34!";

    private FakeButtonService fakeButtonService;
    private UserListView userListView;

    public UserListPresenterImpl(UserListView userListView,
                                 FakeButtonService fakeButtonService) {
        this.userListView = userListView;
        this.fakeButtonService = fakeButtonService;
    }

    @Override
    public void getUsers() {
        fakeButtonService.getUsers(CANDIDATE_ID)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe((userListResponse) -> {
                    userListView.onGetUsersSuccess(userListResponse);
                }, (throwable -> {
                    userListView.onGetUsersFailure(throwable);
                }));
    }
}

File: app/src/main/java/io/jeffchang/buttondemo/ui/userlist/presenter/UserListPresenterImpl.kt
package io.jeffchang.buttondemo.ui.userlist.presenter

import io.jeffchang.buttondemo.network.remote.FakeButtonService
import io.jeffchang.buttondemo.network.remote.FakeButtonService.Companion.CANDIDATE_ID
import io.jeffchang.buttondemo.ui.userlist.view.UserListView
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit

/**
 * Implementation of dog list presenter.
 */

class UserListPresenterImpl(private val userListView: UserListView,
                            private val fakeButtonService: FakeButtonService) : UserListPresenter {

    override fun getUsers() {
        fakeButtonService.getUsers(CANDIDATE_ID)
                .delay(300, TimeUnit.MILLISECONDS)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { userListView.onGetUsersSuccess(it) },
                        { userListView.onGetUsersFailure(it) })
    }
}
